### PR TITLE
Check empty subgraph

### PR DESCRIPTION
**Pull request by cmalinmayor**
# Proposed Change
Fix bug related to empty induced subgraph when computing CTC errors. Also now throws an error in TrackingGraph.get_subgraph() when empty nodes are provided, which I'm not sure is the correct behavior, but an "empty" TrackingGraph is poorly defined (e.g. need start and end frame determined by data).

# Types of Changes
What types of changes does your code introduce? Put an x in the boxes that apply.
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature or enhancement
- [ ] Documentation update
- [x] Tests and benchmarks
- [ ] Maintenance (e.g. dependencies, CI, releases, etc.)

Which topics does your change affect? Put an x in the boxes that apply.
- [ ] Loaders
- [ ] Matchers
- [x] Track Errors
- [ ] Metrics
- [x] Core functionality (e.g. `TrackingGraph`, `run_metrics`, `cli`, etc.)

# Checklist
Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.

- [x] I have read the developer/contributing docs.
- [x] I have added tests that prove that my feature works in various situations or tests the bugfix (if appropriate).
- [x] I have checked that I maintained or improved code coverage.
- [ ] I have checked the benchmarking action to verify that my changes did not adversely affect performance.
- [ ] I have written docstrings and checked that they render correctly in the Read The Docs build (created after the PR is opened).
- [ ] I have updated the general documentation including Metric descriptions and example notebooks if necessary.

### CODE REVIEW

File: src/traccuracy/track_errors/_ctc.py [hunk 1]
```diff
@@ -87,9 +88,11 @@ def get_edge_errors(matched_data: Matched):
...
         get_vertex_errors(matched_data)
 
-    induced_graph = comp_graph.get_subgraph(
-        comp_graph.get_nodes_with_flag(NodeFlag.TRUE_POS)
-    ).graph
+    comp_tp_nodes = comp_graph.get_nodes_with_flag(NodeFlag.TRUE_POS)
+    if len(comp_tp_nodes) == 0:
+        induced_graph = nx.DiGraph()
+    else:
+        induced_graph = comp_graph.get_subgraph(comp_tp_nodes).graph
```
Comment (DragaDoncila) [inline src/traccuracy/track_errors/_ctc.py:95]: Since we're accessing the networkx graph anyway using `.graph`, what do you think about just keeping 

`induced_graph = comp_graph.graph.subgraph(comp_tp_nodes)`

here? This would give us an empty graph for the 0 case without having to do the if statement, and would also avoid the extra overhead of getting the `TrackingGraph` subgraph just to access the nx attribute anyway

File: tests/track_errors/test_ctc_errors.py [hunk 2]
```diff
@@ -134,3 +134,53 @@ def test_assign_edge_errors_semantics():
...
+
+    mapping = [
+        (1, 1),
+        (5, 1),
+        (2, 2),
+        (5, 2),
+    ]
+
+    matched_data = Matched(TrackingGraph(gt), TrackingGraph(comp), mapping)
+    get_edge_errors(matched_data)
```
Comment (DragaDoncila) [inline tests/track_errors/test_ctc_errors.py:186]: would be good here to actually assert on the counts of each error

**Review by DragaDoncila — COMMENTED** [review]
Thanks for fixing this one @cmalinmayor - left just a couple of comments

**Review by msschwartz21 — COMMENTED** [review]
Generally this seems reasonable to me and I'm in agreement with Drag's comments. My only question is if we really want to fail and throw a ValueError in the case where the subgraph is empty. I'm mostly thinking about downstream uses where maybe it would be easier to just return None and then it's the responsibility of the downstream code to handle that case if it appears.

**Comment by DragaDoncila** [comment]
> My only question is if we really want to fail and throw a ValueError in the case where the subgraph is empty.

Yeah I had this thought too actually. I wonder if we should just mimic the networkx behaviour again here, so returning an empty graph. I think that's also the correct subset result for an intersection with an empty set. Based on both of these things it's probably also the "expected" result.

File: src/traccuracy/_tracking_graph.py [hunk 3]
```diff
@@ -316,20 +316,27 @@ def get_connected_components(self) -> list[TrackingGraph]:
...
-            nodes (list): List of node ids to use in constructing the subgraph
-        """
+            nodes (list): A non-empty list of node ids to use in constructing the subgraph
 
+        Raises:
+            ValueError if nodes is empty
+        """
         new_graph = self.graph.subgraph(nodes).copy()
+        logger.debug(f"Subgraph has {new_graph.number_of_nodes()} nodes")
+        if new_graph.number_of_nodes() == 0:
```
[the remark this answers is not in the document]
Comment (cmalinmayor) [reply src/traccuracy/_tracking_graph.py:329]: The only issue is you can techncially pass nodes that aren't in the graph to subgraph, and it will take the intersection of the passed list of nodes and those in the graph. So a len(nodes) != 0 does not necessarily mean the subgraph will have nodes. This may not matter if we allow empty TrackingGraphs as discussed below

**Comment by cmalinmayor** [comment]
@msschwartz21 @DragaDoncila Thanks for the feedback. The main issue with returning an empty TrackingGraph was the start_frame and end_frame attributes, which assume the graph is not empty in our current implementation. After some attempts to fix this by setting them to None if its empty and running into typing issues, I decided to try removing these attributes. Let me know what you think. It means we can create an empty TrackingGraph easily, can avoid checking if the subgraph is empty and just return an empty TrackingGraph, and can get the start_frame and end_frame by sorting `TrackingGraph.nodes_by_frame.keys()` when we need them.

File: src/traccuracy/_tracking_graph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DragaDoncila) [reply src/traccuracy/_tracking_graph.py]: Ah I see - ok let's wait on that discussion and then resolve this.

**Comment by DragaDoncila** [comment]
> The main issue with returning an empty TrackingGraph was the start_frame and end_frame attributes, which assume the graph is not empty in our current implementation. After some attempts to fix this by setting them to None if it's empty and running into typing issues, I decided to try removing these attributes.

Hmm @cmalinmayor I wonder what the danger is of just keeping the `min_frame` and `max_frame` of the original `TrackingGraph`? Does the `segmentation` (if it's there) get given to the subgraph? If so then you could think of this as the same "problem" instance, but just with a different set of detections now (i.e. none) - could be useful?  We already allow frames with no detections, so this would just be the extreme case.

**Comment by cmalinmayor** [comment]
@DragaDoncila I added asserts on the expected behavior in my minimal test case. Turns out, I don't understand the expected behavior 🤷‍♀️ There are two inconsistencies between what I would expect and what happens in the example. 

1. What should happen to the ground truth nodes if the comp nodes are marked Non-split? I know what I did [here](https://github.com/Janelia-Trackathon-2023/traccuracy/blob/bd12224e49c4c8370e1ecbb15b97d5fc448754c0/tests/track_errors/test_ctc_errors.py#L199) was wrong, because the non-split attribute can only be on the computed graph. But when I print all the attributes on the gt nodes for the example:

```
gt         1 - 2 - 3
           4 - 5 - 6
comp       1 - 2
matching [ (1, 1), (4, 1), (2, 2), (5, 2) ]

GT NODES
1 {'t': 0, 'x': 1, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: False}
2 {'t': 1, 'x': 1, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: False}
4 {'t': 0, 'x': 0, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: True} *(huh?)
5 {'t': 1, 'x': 0, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: False}
3 {'t': 2, 'x': 1, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: True} *(expected)
6 {'t': 2, 'x': 0, 'y': 1, <NodeFlag.FALSE_NEG: 'is_fn'>: True} *(expected)
```
One of the four GT nodes matched to a non-split node is a false negative, and the others are not. Which is correct?

2. If the ground truth edges between nodes matched to "non-split" nodes are false negative edges as previously discussed, why are the comp edges not false positives? [This](https://github.com/Janelia-Trackathon-2023/traccuracy/blob/bd12224e49c4c8370e1ecbb15b97d5fc448754c0/tests/track_errors/test_ctc_errors.py#L194) assert fails. All edge attributes printed for the example:
```
gt         1 - 2 - 3
           4 - 5 - 6
comp       1 - 2
matching [ (1, 1), (4, 1), (2, 2), (5, 2) ]

COMP EDGE
(1, 2) {<EdgeFlag.FALSE_POS: 'is_fp'>: False, <EdgeFlag.WRONG_SEMANTIC: 'is_wrong_semantic'>: False, <EdgeFlag.INTERTRACK_EDGE: 'is_intertrack_edge'>: False}
GT EDGES
(1, 2) {<EdgeFlag.FALSE_NEG: 'is_fn'>: True, <EdgeFlag.INTERTRACK_EDGE: 'is_intertrack_edge'>: False}
(2, 3) {<EdgeFlag.FALSE_NEG: 'is_fn'>: True, <EdgeFlag.INTERTRACK_EDGE: 'is_intertrack_edge'>: False}
(4, 5) {<EdgeFlag.FALSE_NEG: 'is_fn'>: True, <EdgeFlag.INTERTRACK_EDGE: 'is_intertrack_edge'>: False}
(5, 6) {<EdgeFlag.FALSE_NEG: 'is_fn'>: True, <EdgeFlag.INTERTRACK_EDGE: 'is_intertrack_edge'>: False}
```

**Comment by cmalinmayor** [comment]
@DragaDoncila @msschwartz21 how do you feel about merging these fixes ASAP even though we have uncovered other questions by adding more thorough tests? I really need the key error and the empty subgraph issues fixed. At Morgan's suggestion, I added back in the start_frame and end_frame and set them to None if the TrackingGraph is empty (both at initialization and subgraph steps).

**Comment by DragaDoncila** [comment]
@cmalinmayor I'm happy to merge the fix even though we've uncovered more questions! Are the currently failing tests a result of the new assertions you tried to make? If so, it would be good to either remove them to get CI green or maybe even better, mark them as `skip` so that we have them but aren't running them and we can just unmark them in the PR that addresses the new inconsistencies you've found.

> What should happen to the ground truth nodes if the comp nodes are marked Non-split? 

According to the paper, the induced subgraph that we compute after we compute the node matching is formed strictly of the "uniquely matching computed vertices and their incident edges", so all GT nodes associated with an NS node should be deleted in the subgraph. This kinda implies that they should **both** be marked FN...?

> If the ground truth edges between nodes matched to "non-split" nodes are false negative edges as previously discussed, why are the comp edges not false positives?

This one is super weird but our induced subgraph in this case would be empty (we have two non-split nodes so they both get removed and by extension their edge also gets removed). The paper says "We define the set of redundant edges (FP edges) in the computed graph as the induced subgraph edges without any counterpart in the ground truth graph", so, because that single edge doesn't exist in the induced subgraph, it doesn't get marked as FP and you just get penalized once for your FN edges from the GT...

**Comment by cmalinmayor** [comment]
> This one is super weird but our induced subgraph in this case would be empty (we have two non-split nodes so they both get removed and by extension their edge also gets removed). The paper says "We define the set of redundant edges (FP edges) in the computed graph as the induced subgraph edges without any counterpart in the ground truth graph", so, because that single edge doesn't exist in the induced subgraph, it doesn't get marked as FP and you just get penalized once for your FN edges from the GT...

Wouldn't all edges adjacent to NS nodes get removed, even if one endpoint is still there? Then we never get penalized for edges next to NS nodes being FP? If that's expected, then I can change the assert for the edges and just kick the can on the node situation.

**Comment by DragaDoncila** [comment]
> Wouldn't all edges adjacent to NS nodes get removed, even if one endpoint is still there? 

Yeah that's correct.

> Then we never get penalized for edges next to NS nodes being FP?

I think that's also correct. If you have an NS node, it gets removed from the induced graph, as do all of its adjacent edges. So I guess for NS nodes you **only** get penalized for their NS weight plus any FN edges that are now created by the fact that you've removed edges adjacent to the NS node.

> If that's expected, then I can change the assert for the edges and just kick the can on the node situation.

I think that sounds ok

File: src/traccuracy/_tracking_graph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmalinmayor) [reply src/traccuracy/_tracking_graph.py]: This check has been removed, since we now allow empty TrackingGraphs

File: tests/track_errors/test_ctc_errors.py (hunk 2 above)
Comment (cmalinmayor) [reply tests/track_errors/test_ctc_errors.py:187]: Let's count this as semi-resolved 😅

File: src/traccuracy/matchers/_ctc.py [hunk 4]
```diff
@@ -51,8 +51,11 @@ def _compute_mapping(self, gt_graph: TrackingGraph, pred_graph: TrackingGraph):
         if mask_gt.shape != mask_pred.shape:
             raise ValueError("Segmentation shapes must match between gt and pred")
 
-        mapping = []
+        mapping: list[tuple] = []
         # Get overlaps for each frame
+        if gt.start_frame is None or gt.end_frame is None:
```
Comment (msschwartz21) [inline src/traccuracy/matchers/_ctc.py:56]: Just checking my understanding, could this also be an `and` because we wouldn't expect the start or end frame to be None without the other also being none?

**Review by msschwartz21 — APPROVED** [review]
Left one question just checking my understanding, but otherwise this looks good to go!

File: src/traccuracy/matchers/_ctc.py (hunk 4 above)
Comment (cmalinmayor) [reply src/traccuracy/matchers/_ctc.py:56]: Right, because they are always set together

File: tests/track_errors/test_ctc_errors.py [hunk 5]
```diff
@@ -134,3 +134,69 @@ def test_assign_edge_errors_semantics():
...
+    matched_data = Matched(TrackingGraph(gt), TrackingGraph(comp), mapping)
+    get_vertex_errors(matched_data)
+    get_edge_errors(matched_data)
+
+    for node in comp.nodes:
+        assert comp.nodes[node][NodeFlag.NON_SPLIT]
+    for edge in comp_edges:
+        assert not comp.edges[edge][EdgeFlag.FALSE_POS]
+
+    if False:  # TODO: Fix this in a separate PR
```
Comment (DragaDoncila) [inline tests/track_errors/test_ctc_errors.py:194]: @cmalinmayor can you link back to the original comment here, just in case we forget what we had to fix

https://github.com/Janelia-Trackathon-2023/traccuracy/pull/141#issuecomment-2265990197

**Review by DragaDoncila — APPROVED** [review]
I haven't pulled this down to play with it as I'm away until next Thursday, but I had a look through and played before so looks good to me, just left a small comment!
